### ISSUE TITLE
Questions on Reproducing Qwen2-VL Zero-Shot Results on M3CoT

### ISSUE THREAD

**Issue by Sioonn** [jungao1106/ICoT#35]
Hi, thank you for sharing this great work!

I've been trying to reproduce the ICoT score of 44.1 for Qwen-VL + M3CoT reported in Table 1. While running the code, I encountered a few points that I'd appreciate clarification on.

<br>

## 1. `max_new_tokens=128` Setting

The generation config in `qwen2_vl/icot_qwen_utils.py` (lines 55–60) is defined as follows:

```python
generation_config = {
    'do_sample': True,
    'temperature': 0.8,
    'top_p': 0.9,
    'max_new_tokens': 128
}
```

When running this on the full M3CoT test set (2,318 samples), a significant number of samples were truncated before generating the final answer (e.g. `"(B)"`), as 128 tokens was not sufficient. Increasing `max_new_tokens` to 512 resolved this, with most samples generating complete answers.

Could you confirm whether 128 was the intended value, or if a different value was used in the paper's experiments?

<br>

## 2. ADS Trigger Rate

Based on the ADS trigger condition in `qwen2_vl/icot_qwen_model.py` (line 134), I observed that ADS was triggered for only about **25.3% (587/2,318)** of the samples across the full M3CoT test set.

My understanding from the paper was that ADS would be triggered for the majority of samples, so the ~25% rate was lower than expected. Could you confirm whether this trigger rate is consistent with your experiments?

<br>

## 3. Newline Token Detection Scope

The newline detection for ADS triggering in `qwen2_vl/icot_qwen_model.py` (line 80) is implemented as:

```python
judge = input_ids[:, -1] in [198]
```

Token ID 198 corresponds to a standalone `'\n'`. However, upon analyzing the actual outputs of Qwen2-VL-7B-Instruct with the tokenizer, I found that the model far more frequently generates **merged tokens** that include a newline combined with preceding punctuation, such as `'.\n'` (624), `':\n'` (510), `':\n\n'` (1447), and `'.\n\n'` (382).

Since these merged tokens have different IDs from 198, they do not trigger ADS under the current implementation. I suspect this may be the cause of the low trigger rate observed in issue 2 above.

Could you clarify whether detecting only the standalone `'\n'` (198) is the intended design, or whether all tokens containing a newline were originally intended to be captured?

<br>
<br>

Thank you!

**Comment by jungao1106** [comment]
Hi, 

1. Regarding experimental configurations, hyperparameters vary by setting, with 512 indeed being utilized where appropriate to ensure optimal performance as you mentioned.

2. As a training-free prompting method, ICoT is designed to be adaptive rather than mandatory for every instance. For simpler queries, the model is permitted to provide direct answers without triggering a full reasoning chain, thereby maintaining efficiency.

3. For the triggering mechanism, any tokenizer element, such as **`\n`**, **`\n\n`**, or variants with spaces, can serve as a trigger, and models can be evaluated using any combination of these. Our experiments, particularly with Qwen models,  indicate that while expanding the set of trigger tokens increases the trigger rate, it also introduces a potential that an excessively high trigger frequency may degrade output quality by causing the model to generate repetitive or meaningless sequences.


Best,
Jun

**Comment by Sioonn** [comment]
Thank you for the clarification!

I have a few additional questions regarding the zero-shot reproduction with Qwen2-VL.

In Table 2, an ablation study comparing ICoT and w/o ADS is conducted for Chameleon-7B in a one-shot setting. Have you also performed a similar ADS ablation for Qwen2-VL?

I ran my own ablation on Qwen2-VL with the following setup. Aside from setting `max_new_tokens=512` as mentioned previously, all other configurations were kept identical to the original. The only modification was adding ADS enable/disable control:

1. Added `getattr(self, 'ads_enabled', True)` to the ADS trigger condition at line 134 of `qwen2_vl/icot_qwen_model.py`:
```python
if getattr(self, 'ads_enabled', True) and self.num_line_break and self.num_line_break % 2 == 0 and self.num_sub_imgs < 3:
```
2. Added the following line to `icot_qwen2vl.py`:
```python
model.ads_enabled = args.ads.lower() == 'true'
```

I then ran two experiments:

```bash
# w/o ADS
python qwen2_vl/icot_qwen2vl.py \
  --ads false \
  --output ./results/qwen2-vl/m3cot_patch/exp1_cot_sample.json \
  > ./results/qwen2-vl/m3cot_patch/exp1.log 2>&1 &

# ICoT (w/ ADS)
python qwen2_vl/icot_qwen2vl.py \
  --ads true \
  --output ./results/qwen2-vl/m3cot_patch/exp2_ads_sample.json \
  > ./results/qwen2-vl/m3cot_patch/exp2.log 2>&1 &
```

The results were as follows:

| Experiment | Accuracy |
|---|---|
| w/o ADS (CoT, sampling) | 45.1% |
| ICoT (ADS, sampling) | 42.2% |

Unlike the paper's reported results where ICoT outperforms w/o ADS, my experiments show the opposite trend. **Have you observed a similar reversal when running Qwen2-VL in the zero-shot setting?**

---

## Further Analysis: Early Termination After ADS Trigger

To investigate this further, I broke down the ICoT results by the number of ADS triggers per sample and compared accuracy and failure rate. The failure rate was measured using the eval_scripts provided in the repository.

| Group | n | Avg Tokens | Accuracy | Failed % |
|---|---|---|---|---|
| w/o ADS (baseline) | 2318 | 223.3 | 45.08% | 1.90% |
| ICoT (trigger=0) | 1731 | 194.0 | 49.45% | 0.92% |
| ICoT (trigger=1) | 100 | 118.8 | 19.00% | 42.00% |
| ICoT (trigger=2) | 31 | 133.7 | 19.35% | 35.48% |
| ICoT (trigger=3) | 456 | 303.8 | 21.49% | 25.88% |

The failure rate spikes sharply as soon as ADS is triggered even once. To understand why, I debugged individual ADS-triggered samples and found that generation frequently terminates prematurely immediately after ADS fires, even when the response is incomplete and well within the `max_new_tokens=512` limit.

For example, running sample `mathematics-1552` from M3CoT with `skip_special_tokens=False` produces:

```
"To determine how long Ted has to wait until the next ferry after missing the 3:15 PM ferry at Seagull Port, we need to look at the ferry schedule for Seagull Port.\n\nThe ferry schedule for Seagull Port is as follows:\n- 3:15 PM\n- 3:30 PM\n\n-<|im_end|>"
```

The generation ends with `<|im_end|>` right after ADS triggers, without reaching a final answer. The per-trigger breakdown above confirms this is not an isolated case but occurs broadly across ADS-triggered samples.

**Have you encountered this early termination behavior when running Qwen2-VL? And do you have any insight into why `<|im_end|>` is being generated immediately after ADS fires — is this an expected interaction between ADS and Qwen2-VL's EOS token behavior, or could it be a bug in the current implementation?**

Thank you in advance for your help.

**Comment by jungao1106** [comment]
Hi, 

Regarding your inquiry, we did not perform an ablation study on Qwen2-VL. The ICoT framework was primarily designed for unimodal VLMs like Chameleon.

Regarding the issue of early termination, we observed similar behavior in our experiments. To resolve this and ensure the model produces a complete reasoning chain, we recommend increasing the min_new_tokens parameter.

Best regards,

Jun
